pass -O options when generating dependency

zasdfgbnm opened the pull request:
In the current implementation, when generating the dependency in [L126](https://github.com/tensorflow/tensorflow/blob/master/third_party/gpus/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc.tpl#L216), the optimization option is not passed to nvcc, which makes nvcc to generate a lot of warnings that looks like:

```
_FORTIFY_SOURCE requires compiling with optimization (-O)
```
This PR clear these warnings by passing optimization option when generating dependency.

The following issues are fixed by this PR: 
https://github.com/tensorflow/tensorflow/issues/9149
https://github.com/tensorflow/tensorflow/issues/2153

tensorflow-jenkins commented:
Can one of the admins verify this patch?

frankchn commented:
Jenkins, test this please.

vrv commented:
@zasdfgbnm can you clarify a bit more about the purpose of what you are doing here?

https://github.com/tensorflow/tensorflow/pull/10999/commits/1bd3bfa8bd9a10c6431d106799d4bada849006fb#diff-04aeac3b8c8dc82fda795b4f8da66489R191 is already setting some options for -O, and 'opt' is passed to the nvcc command line; and we figured it would apply to both the host compiler and nvcc compiler options.  Can you help us out and let us know what specifically is not working as intended, and maybe why too?

zasdfgbnm commented:
@vrv  To be honest, there is also some place that I don't understand well either. 

All I can say is, `crosstool_wrapper_driver_is_not_gcc` tool run commands that looks like

```
PATH=/usr/bin:$PATH /opt/cuda/bin/nvcc -D_FORCE_INLINES -gencode=arch=compute_61,\"code=sm_61,compute_61\" -gencode=arch=compute_35,\"code=sm_35,compute_35\"  --expt-relaxed-constexpr --ftz=true -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNDEBUG -DEIGEN_MPL2_ONLY -DGOOGLE_CUDA=1 -std=c++11 --compiler-options " -isystem external/eigen_archive -isystem bazel-out/local_linux-py3-opt/genfiles/external/eigen_archive -isystem external/bazel_tools/tools/cpp/gcc3 -isystem external/protobuf/src -isystem bazel-out/local_linux-py3-opt/genfiles/external/protobuf/src -isystem external/local_config_cuda/cuda -isystem bazel-out/local_linux-py3-opt/genfiles/external/local_config_cuda/cuda -isystem external/local_config_cuda/cuda/cuda/include -isystem bazel-out/local_linux-py3-opt/genfiles/external/local_config_cuda/cuda/cuda/include -iquote . -iquote bazel-out/local_linux-py3-opt/genfiles -iquote external/eigen_archive -iquote bazel-out/local_linux-py3-opt/genfiles/external/eigen_archive -iquote external/bazel_tools -iquote bazel-out/local_linux-py3-opt/genfiles/external/bazel_tools -iquote external/local_config_sycl -iquote bazel-out/local_linux-py3-opt/genfiles/external/local_config_sycl -iquote external/protobuf -iquote bazel-out/local_linux-py3-opt/genfiles/external/protobuf -iquote external/local_config_cuda -iquote bazel-out/local_linux-py3-opt/genfiles/external/local_config_cuda -g0 -fno-canonical-system-headers -fPIC" --compiler-bindir=/usr/bin/gcc-5 -I . -x cu  -O2 -c tensorflow/contrib/rnn/kernels/gru_ops_gpu.cu.cc -o bazel-out/local_linux-py3-opt/bin/tensorflow/contrib/rnn/_objs/python/ops/_gru_ops_gpu/tensorflow/contrib/rnn/kernels/gru_ops_gpu.cu.pic.o
```

According to my test, if this command is invoked in a shell, it works well. But for some reason (that I don't know) when it is invoked in a bazel build environment, the `-O` option passed to nvcc will not be passed to the host compiler. That's why I make the modification to pass the `-O` option manually to `--compiler-options`. I guess this issue might be caused by some compatibility issue of bazel and nvcc. Can anyone who is familiar with bazel and nvcc give some point?

zasdfgbnm committed:
use optimization when generating dependency

This is to avoid _FORTIFY_SOURCE requires compiling with optimization (-O) warning during compile

zasdfgbnm commented:
@vrv After looking deeper again into the problem, it seems that I "fixed" the problem in a wrong way. I have updated the title, the commits, and description of this pull request. Now the changes are straightforward I think. Thank you very much for the hint.

@keveman Hopefully this PR should be a small change and can be easy to review now.

vrv commented:
This makes a lot more sense, thank you for digging deeper! :)

@tensorflow-jenkins test this please

zasdfgbnm commented:
Seems to be unrelated fails?
